src/porepy/examples/fracture_damage.py

```diff
@@ -0,0 +1,446 @@
+    pp.constitutive_laws.FrictionDamage,
+    pp.constitutive_laws.DilationDamage,
+    damage.DamageHistoryVariable,
+    damage.DamageHistoryEquation,
+    pp.contact_mechanics.ContactMechanics,
+):
+    """Fracture damage model.
+
+    This class needs to be combined with a class defining the history equation, i.e.,
+    (An)IsotropicHistoryEquation. This is to allow for different history equations to be
+    used with the same model, specifically during testing. TODO: Consider to choose one
+    as default.
+
+    Two methods are overridden in this class: - variables_stored_all_time_steps: The
+    interface displacement is not included in the
+      model, and thus the method is overridden to exclude it.
+    - update_time_dependent_interface_displacement: The interface displacement parameter
+      needs to be stored at all time steps.
+
+    The combination of the two overrides amounts to replacing the variable stored at
+    each time step with the parameter value being stored.
+
+    """
+
+    def variables_stored_all_time_steps(self) -> list[pp.ad.Variable]:
+        """Return the variables stored at all time steps.
+
+        Override default implementation, since the interface displacement is not
+        included in the model.
+
+        Returns:
+            List of variables.
+
+        """
+        return self.equation_system.get_variables(
+            variables=[
+                self.contact_traction_variable,
+            ]
+        )
+
+    def update_time_dependent_interface_displacement(self) -> None:
+        """Update the interface displacement parameter."""
+
+        name = self.interface_displacement_parameter_key
+        for intf, data in self.mdg.interfaces(return_data=True):
+            if pp.ITERATE_SOLUTIONS in data and name in data[pp.ITERATE_SOLUTIONS]:
+                # Use the values at the unknown time step from the previous time step.
+                vals = pp.get_solution_values(name=name, data=data, iterate_index=0)
+            else:
+                # No current value stored. The method was called during the
+                # initialization.
+                vals = self.interface_displacement_parameter_values(intf).ravel(
+                    order="F"
+                )
+
+            # Before setting the new, most recent time step, shift the stored values
+            # backwards in time.
+            pp.shift_solution_values(
+                name=name,
+                data=data,
+                location=pp.TIME_STEP_SOLUTIONS,
+                max_index=None,
+            )
+            # Set the values of current time to most recent previous time.
+            pp.set_solution_values(name=name, values=vals, data=data, time_step_index=0)
+
+            # Set the unknown time step values.
+            vals = self.interface_displacement_parameter_values(intf).ravel(order="F")
+            pp.set_solution_values(name=name, values=vals, data=data, iterate_index=0)
+
+
+@dataclass
```

Comment:
EK note to self: I stopped reading here.

Reply:
Can this be resolved?